File: apps/api/src/veupath_chatbot/services/experiment/types/experiment.py
"""Experiment and ExperimentConfig types."""

from pydantic import Field

from veupath_chatbot.domain.strategy.ast import PlanStepNode
from veupath_chatbot.platform.pydantic_base import CamelModel, RoundedFloat2
from veupath_chatbot.platform.types import JSONObject
from veupath_chatbot.services.experiment.types.core import (
    DEFAULT_STEP_ANALYSIS_PHASES,
    ControlValueFormat,
    EnrichmentAnalysisType,
    ExperimentMode,
    ExperimentStatus,
    OptimizationObjective,
)
from veupath_chatbot.services.experiment.types.enrichment import EnrichmentResult
from veupath_chatbot.services.experiment.types.metrics import (
    CrossValidationResult,
    ExperimentMetrics,
    GeneInfo,
)
from veupath_chatbot.services.experiment.types.optimization import (
    OperatorKnob,
    OptimizationSpec,
    ThresholdKnob,
    TreeOptimizationResult,
)
from veupath_chatbot.services.experiment.types.rank import (
    BootstrapResult,
    RankMetrics,
)
from veupath_chatbot.services.experiment.types.step_analysis import StepAnalysisResult


class ExperimentConfig(CamelModel):
    """Full configuration for an experiment run.

    Supports three modes:

    * **single** (default): one search + parameters.
    * **multi-step**: a recursive ``step_tree`` of search/combine/transform nodes.
    * **import**: import an existing Pathfinder strategy by ``source_strategy_id``.
    """

    site_id: str
    record_type: str
    search_name: str
    parameters: JSONObject
    positive_controls: list[str]
    negative_controls: list[str]
    controls_search_name: str
    controls_param_name: str
    controls_value_format: ControlValueFormat = "newline"
    enable_cross_validation: bool = False
    k_folds: int = 5
    enrichment_types: list[EnrichmentAnalysisType] = Field(default_factory=list)
    name: str = ""
    description: str = ""
    optimization_specs: list[OptimizationSpec] | None = None
    optimization_budget: int = 30
    optimization_objective: OptimizationObjective = "balanced_accuracy"
    parameter_display_values: dict[str, str] | None = None
    mode: ExperimentMode = "single"
    step_tree: PlanStepNode | None = None
    source_strategy_id: str | None = None
    optimization_target_step: str | None = None
    enable_step_analysis: bool = False
    step_analysis_phases: list[str] = Field(
        default_factory=lambda: list(DEFAULT_STEP_ANALYSIS_PHASES)
    )
    control_set_id: str | None = None
    threshold_knobs: list[ThresholdKnob] | None = None
    operator_knobs: list[OperatorKnob] | None = None
    tree_optimization_objective: str = "precision_at_50"
    tree_optimization_budget: int = 50
    max_list_size: int | None = None
    sort_attribute: str | None = None
    sort_direction: str = "ASC"
    parent_experiment_id: str | None = None
    target_gene_ids: list[str] | None = None

    @property
    def is_tree_mode(self) -> bool:
        """Whether this config uses a multi-step strategy tree."""
        return self.mode in ("multi-step", "import") and self.step_tree is not None


class BatchOrganismTarget(CamelModel):
    """Per-organism overrides for a cross-organism batch experiment."""

    organism: str
    positive_controls: list[str] | None = None
    negative_controls: list[str] | None = None


class BatchExperimentConfig(CamelModel):
    """Configuration for running the same search across multiple organisms."""

    base_config: ExperimentConfig
    organism_param_name: str
    target_organisms: list[BatchOrganismTarget] = Field(default_factory=list)


class Experiment(CamelModel):
    """Full experiment with config and results."""

    id: str
    config: ExperimentConfig
    user_id: str | None = None
    status: ExperimentStatus = "pending"
    metrics: ExperimentMetrics | None = None
    cross_validation: CrossValidationResult | None = None
    enrichment_results: list[EnrichmentResult] = Field(default_factory=list)
    true_positive_genes: list[GeneInfo] = Field(default_factory=list)
    false_negative_genes: list[GeneInfo] = Field(default_factory=list)
    false_positive_genes: list[GeneInfo] = Field(default_factory=list)
    true_negative_genes: list[GeneInfo] = Field(default_factory=list)
    error: str | None = None
    total_time_seconds: RoundedFloat2 | None = None
    created_at: str = ""
    completed_at: str | None = None
    batch_id: str | None = None
    benchmark_id: str | None = None
    control_set_label: str | None = None
    is_primary_benchmark: bool = False
    optimization_result: JSONObject | None = None
    wdk_strategy_id: int | None = None
    wdk_step_id: int | None = None
    notes: str | None = None
    step_analysis: StepAnalysisResult | None = None
    rank_metrics: RankMetrics | None = None
    robustness: BootstrapResult | None = None
    tree_optimization: TreeOptimizationResult | None = None

    def classification_id_sets(
        self,
    ) -> tuple[set[str], set[str], set[str], set[str]]:
        """Build classification ID sets from the gene lists.

        :returns: ``(tp_ids, fp_ids, fn_ids, tn_ids)``
        """
        return (
            {g.id for g in self.true_positive_genes},
            {g.id for g in self.false_positive_genes},
            {g.id for g in self.false_negative_genes},
            {g.id for g in self.true_negative_genes},
        )

    def result_gene_ids(self) -> set[str]:
        """Return the set of all result gene IDs (TP + FP)."""
        return {g.id for g in self.true_positive_genes} | {
            g.id for g in self.false_positive_genes
        }

File: apps/web/src/lib/api/schemas/experiment.ts
/**
 * Zod schemas for Experiment API responses.
 *
 * Summary schemas match the flat shape from experiment_summary_to_json().
 * Detail schemas match the nested shape from experiment_to_json() (model_dump).
 */
import { z } from "zod";
import { DateTimeString } from "./common";

// ---------------------------------------------------------------------------
// Shared enums
// ---------------------------------------------------------------------------

const ExperimentStatusSchema = z.enum([
  "pending",
  "running",
  "completed",
  "error",
  "cancelled",
]);

const ExperimentModeSchema = z.enum(["single", "multi-step", "import"]);

// ---------------------------------------------------------------------------
// Experiment Summary (flat — from list endpoint)
// ---------------------------------------------------------------------------

export const ExperimentSummarySchema = z.object({
  id: z.string(),
  name: z.string(),
  siteId: z.string(),
  searchName: z.string(),
  recordType: z.string(),
  mode: ExperimentModeSchema,
  status: ExperimentStatusSchema,
  f1Score: z.number().nullable(),
  sensitivity: z.number().nullable(),
  specificity: z.number().nullable(),
  totalPositives: z.number(),
  totalNegatives: z.number(),
  createdAt: DateTimeString,
  batchId: z.string().nullable().optional(),
  benchmarkId: z.string().nullable().optional(),
  controlSetLabel: z.string().nullable().optional(),
  isPrimaryBenchmark: z.boolean(),
});

export const ExperimentSummaryListSchema = z.array(ExperimentSummarySchema);

// ---------------------------------------------------------------------------
// Full Experiment (nested — from detail endpoint)
// ---------------------------------------------------------------------------

// -- Config (nested under experiment.config) --------------------------------

const ExperimentConfigSchema = z.object({
  siteId: z.string(),
  recordType: z.string(),
  searchName: z.string(),
  parameters: z.record(z.string(), z.unknown()),
  positiveControls: z.array(z.string()),
  negativeControls: z.array(z.string()),
  controlsSearchName: z.string(),
  controlsParamName: z.string(),
  controlsValueFormat: z.enum(["newline", "json_list", "comma"]),
  enableCrossValidation: z.boolean(),
  kFolds: z.number(),
  enrichmentTypes: z.array(z.string()).optional(),
  name: z.string(),
  description: z.string(),
  mode: ExperimentModeSchema,
  optimizationSpecs: z.array(z.record(z.string(), z.unknown())).nullable().optional(),
  optimizationBudget: z.number().optional(),
  optimizationObjective: z.string().optional(),
  parameterDisplayValues: z.record(z.string(), z.string()).nullable().optional(),
  stepTree: z.unknown().optional(),
  sourceStrategyId: z.string().nullable().optional(),
  optimizationTargetStep: z.string().nullable().optional(),
  enableStepAnalysis: z.boolean().optional(),
  stepAnalysisPhases: z.array(z.string()).optional(),
  controlSetId: z.string().nullable().optional(),
  thresholdKnobs: z.array(z.record(z.string(), z.unknown())).nullable().optional(),
  operatorKnobs: z.array(z.record(z.string(), z.unknown())).nullable().optional(),
  treeOptimizationObjective: z.string().optional(),
  treeOptimizationBudget: z.number().optional(),
  maxListSize: z.number().nullable().optional(),
  sortAttribute: z.string().nullable().optional(),
  sortDirection: z.string().optional(),
  parentExperimentId: z.string().nullable().optional(),
  targetGeneIds: z.array(z.string()).nullable().optional(),
});

// -- Metrics (nested under experiment.metrics) ------------------------------

export const ConfusionMatrixSchema = z.object({
  truePositives: z.number(),
  falsePositives: z.number(),
  trueNegatives: z.number(),
  falseNegatives: z.number(),
});

export const ExperimentMetricsSchema = z.object({
  confusionMatrix: ConfusionMatrixSchema,
  sensitivity: z.number(),
  specificity: z.number(),
  precision: z.number(),
  f1Score: z.number(),
  mcc: z.number(),
  balancedAccuracy: z.number(),
  negativePredictiveValue: z.number(),
  falsePositiveRate: z.number(),
  falseNegativeRate: z.number(),
  youdensJ: z.number(),
  totalResults: z.number(),
  totalPositives: z.number(),
  totalNegatives: z.number(),
});

// -- Gene info --------------------------------------------------------------

const GeneInfoSchema = z.object({
  id: z.string(),
  name: z.string().nullable().optional(),
  organism: z.string().nullable().optional(),
  product: z.string().nullable().optional(),
});

// -- Enrichment -------------------------------------------------------------

const EnrichmentTermSchema = z.object({
  termId: z.string(),
  termName: z.string(),
  geneCount: z.number(),
  backgroundCount: z.number(),
  foldEnrichment: z.number(),
  oddsRatio: z.number(),
  pValue: z.number(),
  fdr: z.number(),
  bonferroni: z.number(),
  genes: z.array(z.string()).optional(),
});

const EnrichmentResultSchema = z.object({
  analysisType: z.string(),
  terms: z.array(EnrichmentTermSchema),
  totalGenesAnalyzed: z.number().default(0),
  backgroundSize: z.number().default(0),
  error: z.string().nullable().optional(),
});

// -- Cross-validation -------------------------------------------------------

const FoldMetricsSchema = z.object({
  foldIndex: z.number(),
  metrics: ExperimentMetricsSchema,
  positiveControlIds: z.array(z.string()).optional(),
  negativeControlIds: z.array(z.string()).optional(),
});

const CrossValidationResultSchema = z.object({
  k: z.number(),
  folds: z.array(FoldMetricsSchema),
  meanMetrics: ExperimentMetricsSchema,
  stdMetrics: z.record(z.string(), z.number()).optional(),
  overfittingScore: z.number().optional(),
  overfittingLevel: z.string().optional(),
});

// -- Rank metrics -----------------------------------------------------------

const RankMetricsSchema = z.object({
  precisionAtK: z.record(z.string(), z.number()).optional(),
  recallAtK: z.record(z.string(), z.number()).optional(),
  enrichmentAtK: z.record(z.string(), z.number()).optional(),
  prCurve: z.array(z.tuple([z.number(), z.number()])).optional(),
  listSizeVsRecall: z.array(z.tuple([z.number(), z.number()])).optional(),
  totalResults: z.number().optional(),
});

// -- Step analysis ----------------------------------------------------------

const StepAnalysisResultSchema = z.object({
  stepEvaluations: z.array(z.record(z.string(), z.unknown())).optional(),
  operatorComparisons: z.array(z.record(z.string(), z.unknown())).optional(),
  stepContributions: z.array(z.record(z.string(), z.unknown())).optional(),
  parameterSensitivities: z.array(z.record(z.string(), z.unknown())).optional(),
});

// -- Bootstrap/robustness ---------------------------------------------------

const BootstrapResultSchema = z.object({
  nIterations: z.number().optional(),
  metricCis: z.record(z.string(), z.record(z.string(), z.unknown())).optional(),
  rankMetricCis: z.record(z.string(), z.record(z.string(), z.unknown())).optional(),
  topKStability: z.number().optional(),
  negativeSetSensitivity: z.array(z.record(z.string(), z.unknown())).optional(),
});

// -- Tree optimization ------------------------------------------------------

const TreeOptimizationResultSchema = z.object({
  bestTrial: z.record(z.string(), z.unknown()).nullable().optional(),
  allTrials: z.array(z.record(z.string(), z.unknown())).optional(),
  totalTimeSeconds: z.number().optional(),
  objective: z.string().optional(),
});

// -- Full Experiment --------------------------------------------------------

export const ExperimentSchema = z.object({
  id: z.string(),
  config: ExperimentConfigSchema,
  userId: z.string().nullable().optional(),
  status: ExperimentStatusSchema,
  metrics: ExperimentMetricsSchema.nullable().optional(),
  crossValidation: CrossValidationResultSchema.nullable().optional(),
  enrichmentResults: z.array(EnrichmentResultSchema).optional(),
  truePositiveGenes: z.array(GeneInfoSchema).optional(),
  falseNegativeGenes: z.array(GeneInfoSchema).optional(),
  falsePositiveGenes: z.array(GeneInfoSchema).optional(),
  trueNegativeGenes: z.array(GeneInfoSchema).optional(),
  error: z.string().nullable().optional(),
  totalTimeSeconds: z.number().nullable().optional(),
  createdAt: DateTimeString,
  completedAt: DateTimeString.nullable().optional(),
  batchId: z.string().nullable().optional(),
  benchmarkId: z.string().nullable().optional(),
  controlSetLabel: z.string().nullable().optional(),
  isPrimaryBenchmark: z.boolean(),
  optimizationResult: z.record(z.string(), z.unknown()).nullable().optional(),
  wdkStrategyId: z.number().nullable().optional(),
  wdkStepId: z.number().nullable().optional(),
  notes: z.string().nullable().optional(),
  stepAnalysis: StepAnalysisResultSchema.nullable().optional(),
  rankMetrics: RankMetricsSchema.nullable().optional(),
  robustness: BootstrapResultSchema.nullable().optional(),
  treeOptimization: TreeOptimizationResultSchema.nullable().optional(),
});
